Convert "Reasoning with inductive types" to prodn, part 2

jfehrle opened the pull request:
For review of changes in `reasoning-inductives.rst` up to ~line 500 ("Equality of inductive types").  Whenever you're ready.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,44 +237,54 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
 
-.. tacn:: induction @term
-   :name: induction
+.. tacn:: induction {+, @induction_clause } {? @induction_principle }
```
I think it would make sense to move the definition of `induction_principle` here.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,44 +237,54 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
-   + The argument :n:`@term` can also be a pattern of which holes are denoted
+   + The argument :n:`@one_term` can also be a pattern in which holes are denoted
      by “_”. In this case, the tactic checks that all subterms matching the
      pattern in the conclusion and the hypotheses are compatible and
```
To be changed like in the case of `destruct`. It doesn't do this check. It just finds the first pattern that matches. As in this example:

```coq
Goal forall n m, n + m = m + n.
intros.
induction _.
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,44 +237,54 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
+   is **IH**.  The tactic is similar to :tacn:`destruct`, except that
+   `destruct` doesn't generate induction hypotheses.
+
+   There are several cases:
```
Note that none of these is specific to `induction` (this is a repetition of what is described in `destruct` so I don't know if we want to keep it). BTW, I like this description with bullet points more than the equivalent documentation scattered around for `destruct` so I would even suggest using this form above.

jfehrle replied:
When the behavior is identical, it's probably better not to repeat it.  So I dropped the text of your previous comment.  If the others were dropped, you would need to mentally substitute "induction" for "destruct" in most of them, so I thought it better to repeat them.  I could add a sentence here before the points saying they're repetitions but with "induction" instead of "destruct".  destruct doesn't describe the induction principle, so that should stay.

I agree that bullet points look nicer, but the other format makes it much easier to scan for elements of the syntax.  I suppose you could have bullet points with the form `. <syntax>: text`, but not sure that would be an improvement.  Would you want to do this change pervasively (and when) or just for some tactics/commands?

Your thoughts?

jfehrle replied:
OTOH, if "induction" is a very commonly used tactic, that would be a small point in favor of repetition.

Zimmi48 replied:
Yes, `induction` is a very commonly used tactic, but so is `destruct` and I prefer having a clear link to `destruct` for the description of how it treats its argument rather than a repetition.

No, I don't want to pervasively go for bullet points in the manual. I quite like the way that we have decided to document syntax elements. However, this case is a bit different IMHO because the bullet points represent *alternatives* and some alternatives are not even part of the cases that one can see from the syntax (`ident`, which is a sub-case of `one_term`).

So what I suggested was to reuse the way it was in the description of `induction`, i.e. first the alternative ways of providing the main argument (with bullet points) then the other syntax elements (with the usual style for describing those).

jfehrle replied:
OK, we're agreed on the style, but I'm still not sure what change you want me to make. Do you want me to copy the bullet points to `destruct` with substitution of `destruct` as needed?  Or what?

Zimmi48 replied:
> Do you want me to copy the bullet points to `destruct` with substitution of `destruct` as needed?

Yes, something like that (+ the clean-up of the redundant info that this would create).

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,44 +237,54 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
      performs induction using this subterm.
 
+   :n:`{+, @induction_clause }`
+     If no using clause is provided, this is equivalent to doing :n:`induction` on the
```
```suggestion
     If no :n:`@induction_principle` is provided, this is equivalent to doing :n:`induction` on the
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,44 +237,54 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
+     If no using clause is provided, this is equivalent to doing :n:`induction` on the
+     first :n:`@induction_clause` followed by :n:`destruct` on any subsequent clauses.
+
+   :n:`using @one_term {? with @bindings }`
```
```suggestion
   :n:`@induction_principle`
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
-   instance, for an inductive type with  one constructor, the pattern notation
-   :n:`(p`:sub:`1` :n:`, ... , p`:sub:`n` :n:`)` can be used instead of
-   :n:`[ p`:sub:`1` :n:`... p`:sub:`n` :n:`]`.
+   .. exn:: Not an inductive product.
```
Move the error messages up in the description of `induction`.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
 
-   This behaves like :tacn:`induction` providing explicit instances for the
-   premises of the type of :n:`term` (see :ref:`bindings`).
+      Use the variant :tacn:`elim` `… with …` instead.
```
Should recommend using the `with @bindings` clause or the `einduction` variant (but not mention `elim`).

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
-   :undocumented:
-
-.. exn:: Unable to find an instance for the variables @ident ... @ident.
+.. example:: :tacn:`induction` with :n:`@occurrences`
```
That's not a very convincing example since it doesn't behave differently than without the `@occurrences` clause.

Here is an example where it does behave differently:

```coq
Lemma induction_test : forall n:nat, n = n -> n <= n.
intros.
induction n in H |-.
Show 2.
```

However, I have no idea if there exists a *useful* example of this clause with `induction`.

jfehrle replied:
If it's not useful then maybe drop the example and also consider whether to drop `occurrences` from the syntax.

Zimmi48 replied:
> also consider whether to drop `occurrences` from the syntax

I guess that it should be kept for uniformity with `destruct` (where it can be more clearly useful). And there might be use cases that I'm neglecting.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
 
-.. tacv:: einduction @term
-   :name: einduction
+.. tacn:: einduction {+, @induction_clause } {? @induction_principle }
```
Much like in the case analysis section, `einduction` should be indented to appear as a variant of `induction` and `ecase` should appear as a variant of `case`.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
-.. tacv:: induction @term with @bindings as @or_and_intropattern using @term with @bindings in @goal_occurrences
-          einduction @term with @bindings as @or_and_intropattern using @term with @bindings in @goal_occurrences
+   An older, more basic induction tactic.  We recommend using :tacn:`induction`
+   instead where possible.  :n:`@one_term` must have an inductive
```
Interesting, that the argument of `induction` must have an inductive type was forgotten in the description of `induction`.

jfehrle replied:
I slipped it into the first bullet point.  Perhaps it should also be stated in "destruct"?

Zimmi48 replied:
It already appears in the description of `destruct`, although maybe not as clearly as it should:

>the inductive or co-inductive type in the one_term in induction_arg

"in the one_term" doesn't make it sufficiently clear that the type of one_term is the inductive (or co-inductive) type.

jfehrle replied:
A couple questions:
- For `A -> B -> C`, is it correct to say that A and B are premises and C is the conclusion?  This would be a second sense for "conclusion", which is currently defined as the statement to be proven.  Maybe say "the conclusion of a type" to clarify?
- Is it worth making "premise" a defined term? It seems fairly generic.  We could also say A and B are "antecedents".  That terminology is used twice in the doc with that meaning.
- Is there a proper term for `A -> B`?  An implication?

Zimmi48 replied:
>For `A -> B -> C`, is it correct to say that A and B are premises and C is the conclusion?  This would be a second sense for "conclusion", which is currently defined as the statement to be proven.  Maybe say "the conclusion of a type" to clarify?

Yes, this is correct. And yes, "conclusion" has this double meaning. I've signalled this when we've documented the other meaning (I was worried about the confusion that this would create).

>Is it worth making "premise" a defined term? It seems fairly generic. We could also say A and B are "antecedents". That terminology is used twice in the doc with that meaning.

Yes, I guess it would be worth (in the place where function types and dependent / non-dependent premises are explained). I am not used to seeing the terminology "antecedent" in this case, so I would avoid it.

>Is there a proper term for `A -> B`?  An implication?

Yes, an "implication" (with the propositional view) or a non-dependent function type (with the programming language / computational view).

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
-          einduction @term with @bindings as @or_and_intropattern using @term with @bindings in @goal_occurrences
+   An older, more basic induction tactic.  We recommend using :tacn:`induction`
+   instead where possible.  :n:`@one_term` must have an inductive
+   type.  The tactic applies the appropriate :term:`induction principle` based
```
The fact that the tactic selects the "appropriate" induction principle (among the various ones that are automatically generated) when the `@induction_principle` clause is not provided is something else that could be added to the description of `induction`.

jfehrle replied:
Done.  Would it be better or worse to say "based on the _sort_ of the goal"?

Zimmi48 replied:
It would be better indeed.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
-   induction loading into the context of hypotheses. More generally,
-   :n:`elim @term` also works when the type of :n:`@term` is a statement
+   More generally,
+   :n:`elim @one_term` also works when the type of :n:`@one_term` is a statement
```
This paragraph also entirely applies to `induction` (and was forgotten from its documentation).

jfehrle replied:
Forgotten?  Did I ever know in the first place :-)?

I copied the paragraph, but I think it overlaps a bit with the text in the preceeding paragraph: "and induction hypotheses are added to the local context using names whose prefix is **IH**"

Zimmi48 replied:
>Forgotten? Did I ever know in the first place :-)?

Forgotten but not by you. Before this PR.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
+   contains :g:`n : nat` and the current goal is :g:`T` of type :g:`Prop`, then
+   :n:`elim k` is equivalent to :tacn:`apply` :n:`nat_ind with (n:=k)` since
+   `nat_ind` is for `Prop` goals.)
+   ``elim`` only modifies the goal; it does not modify the :term:`local context`.
```
That's the main difference between `induction` and `elim` so this should be the thing to put forward in its documentation.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
 
-.. tacv:: elim @term with @bindings
-   :name: elim … with
+  :n:`with @bindings`
```
```suggestion
  :n:`with @bindings` (in :n:`@one_term_with_bindings`)
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
+   Conversely, if :g:`t` is a :n:`@one_term` of (inductive) type :g:`I` that does
    not occur in the goal, then :n:`elim t` is equivalent to
-   :n:`elimtype I; 2:exact t.`
+   :n:`elimtype I; 2:` :tacn:`exact` `t.`
```
The syntax is incorrect. Here's a fix:

```suggestion
   :n:`elimtype I; only 2:` :tacn:`exact` `t.`
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
-   :name: dependent induction
-
-   The *experimental* tactic dependent induction performs induction-
+   The *experimental* tactic :tacn:`dependent induction` performs induction-
```
There shouldn't be a newline between `induction-` and `inversion`. It introduces a spurious space.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -284,155 +294,81 @@ Induction
+   The *experimental* tactic :tacn:`dependent induction` performs induction-
    inversion on an instantiated inductive predicate. One needs to first
-   require the Coq.Program.Equality module to use this tactic. The tactic
+   :cmd:`Require` the Coq.Program.Equality module to use this tactic. The tactic
```
```suggestion
   :cmd:`Require` the `Coq.Program.Equality` module to use this tactic. The tactic
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -442,6 +378,15 @@ Induction
+   There is a long example of :tacn:`dependent induction` and an explanation
+   of the underlying technique :ref:`here <dependent-induction-examples>`.
+
 .. example::
```
It would probably make sense to indent this example (and the `.. seealso`) to make them part of the description of `dependent induction`. Same thing above with the examples about `induction`.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -476,106 +421,172 @@ Induction
 
    This technique works with any inductive predicate. In fact, the
-   ``dependent induction`` tactic is just a wrapper around the ``induction``
+   :tacn:`dependent induction` tactic is just a wrapper around the ``induction``
```
```suggestion
   :tacn:`dependent induction` tactic is just a wrapper around the :tacn:`induction`
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -476,106 +421,172 @@ Induction
 
-.. exn:: No primitive equality found.
-   :undocumented:
+   .. todo:
```
What is this?

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -476,106 +421,172 @@ Induction
+   :n:`with {+ ( @ident {* @simple_binder } {? %{ struct @name %} } : @type ) }`
+     Starts a proof by mutual induction. The statements to be proven
+     are :n:`forall @simple_binder__i, @type__i`.
+     The identifiers :n:`@ident` are the names of the induction hypotheses. The identifiers
```
```suggestion
     The identifiers :n:`@ident` (including the first one before the `with` clause) are the names of the induction hypotheses. The identifiers
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -476,106 +421,172 @@ Induction
+     Starts a proof by mutual induction. The statements to be proven
+     are :n:`forall @simple_binder__i, @type__i`.
+     The identifiers :n:`@ident` are the names of the induction hypotheses. The identifiers
+     :n:`@ident` are the respective names of the premises on which the induction
```
```suggestion
     :n:`@name` are the respective names of the premises on which the induction
```

(there used to be prime symbols to distinguish the two sets of `ident` that were lost during the Sphinx translation, see https://coq.inria.fr/distrib/V8.7.2/refman/tactics.html#hevea_tactic127).

jfehrle replied:
OK?

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -476,106 +421,172 @@ Induction
+   :n:`with {+ ( @ident {* @simple_binder } : @type ) }`
+     Starts a proof by mutual co-induction. The statements to be
+     proven are :n:`forall @simple_binder__i, @type__i`.
+     The identifiers :n:`@ident` are the names of the co-induction hypotheses.
```
```suggestion
     The identifiers :n:`@ident` (including the first one before the `with` clause) are the names of the co-induction hypotheses.
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -476,106 +421,172 @@ Induction
 
-.. exn:: Not a discriminable equality.
-   :undocumented:
+   Starts a proof by co-induction. The first :n:`@ident` is the
```
This "first" word is problematic since all the `ident` are the names of co-induction hypotheses but only the first one is provided if there is no `with` clause. So either use plural from the beginning, or apply this suggestion:

```suggestion
   Starts a proof by co-induction. The identifier :n:`@ident` is the
```

jfehrle replied:
OK?

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -476,106 +421,172 @@ Induction
+   verification that the use of co-inductive hypotheses is correct is
+   done only at the time of registering the lemma in the global environment. To
+   know if the use of co-induction hypotheses is correct at some time of
+   the interactive development of a proof, use the command ``Guarded``
+   (see Section :ref:`requestinginformation`).
```
```suggestion
   the interactive development of a proof, use the command :cmd:`Guarded`.
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -476,106 +421,172 @@ Induction
+   arguments are correct is done only when registering the
+   lemma in the global environment. To know if the use of induction hypotheses
+   is correct during the interactive development of a proof, use
+   the command :cmd:`Guarded` (see Section :ref:`requestinginformation`).
```
```suggestion
   the command :cmd:`Guarded`.
```

Zimmi48 commented:
> For review of changes in `reasoning-inductives.rst` up to ~line 500 ("Equality of inductive types").

Done.

jfehrle commented:
Updated.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -249,83 +253,93 @@ Induction
+   which the induction principle applies and otherwise fails.
+
+   Most parameters behave as they do in :tacn:`destruct`, with the following
+   differences:
```
There might be a misunderstanding. What is described below do not represent differences with `destruct`. On the contrary, `destruct` interprets its main argument in the same way.

jfehrle replied:
I struck out the part about differences.  Are we good now?

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,202 +236,153 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
-   the unresolved premises are posed as existential variables to be inferred
-   later, in the same way as :tacn:`eapply` does.
+   :n:`{+, @induction_clause }`
+     If no :n:`induction_principle` clause is provided, this is equivalent to doing
```
```suggestion
     If no :n:`@induction_principle` clause is provided, this is equivalent to doing
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,202 +236,153 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
-.. tacv:: induction @term using @term
-   :name: induction … using …
+   :n:`@induction_principle`
+     :n:`one_term` specifies which :term:`induction principle` to use and, optionally,
```
```suggestion
     :n:`@one_term` specifies which :term:`induction principle` to use and, optionally,
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -249,83 +253,93 @@ Induction
-   instead where possible.  :n:`@one_term` must have an inductive
+   An older, more basic induction tactic.  Unlike :tacn:`induction`, ``elim`` only
+   modifies the goal; it does not modify the :term:`local context`.  We recommend
+   using :tacn:`induction` instead where possible.  :n:`@one_term` must have an inductive
```
Since everything else also applies to `induction`, the main description of `elim` could really stop after this second sentence.

jfehrle replied:
I trimmed only this paragraph and kept "We recommend using induction...".  Did you want me to drop more here (e.g. up to `tacn:: eelim`)?

Zimmi48 replied:
You can also remove the paragraph:

>More generally, elim [...] and otherwise fails.

Since it's a repetition of what is (now) said for `induction`.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,202 +236,153 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
 
-.. tacv:: eelim @term
-   :name: eelim
+.. tacn:: eelim @one_term_with_bindings {? using @one_term {? with @bindings } }
```
Don't forget to indent `eelim`.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -452,33 +465,34 @@ Induction
-
-   Starts a proof by co-induction. The first :n:`@ident` is the
-   name of the co-induction hypothesis. As in a cofix expression,
+   Starts a proof by co-induction. The :n:`@ident`\s (including the first one
```
There's a lack of consistency in the description of `fix` and `cofix`. Either the main description should focus on the case without the `with` clause and `ident` should be then singular (in both cases) or it should already encompass the optional use of the `with` clause and then it should be plural in both cases (possibly with the `s` in parentheses).

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -442,140 +392,215 @@ Induction
+   verification that the use of co-inductive hypotheses is correct is
+   done only at the time of registering the lemma in the global environment. To
+   know if the use of co-induction hypotheses is correct at some time of
+   the interactive development of a proof, use the command :cmd`Guarded`.
```
```suggestion
   the interactive development of a proof, use the command :cmd:`Guarded`.
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -442,140 +392,215 @@ Induction
+     are :n:`forall @simple_binder__i, @type__i`.
+     The identifiers :n:`@ident` (including the first one before the `with` clause)
+     are the names of the induction hypotheses. The identifiers
+     :n:`@name` are the respective names of the premises on which the induction
```
We can maybe clarify even further like this:
```suggestion
     :n:`@name` (in the `{ struct ... }` clauses) are the respective names of the premises on which the induction
```

jfehrle commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,202 +236,153 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
+
+   + If :n:`@one_term` (in :n:`@induction_arg`)
+     is an identifier :n:`@ident` denoting a quantified variable of the
+     conclusion of the goal, then :n:`induction @ident` behaves like :tacn:`intros` :n:`until
```
e.g., refers to `B` when the goal is `A -> B`?

Zimmi48 replied:
In this case, I'm afraid that's the other meaning, i.e. the part below the `--------`.

jfehrle commented:
Updated.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -119,16 +119,17 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
       | @one_term_with_bindings
-      induction_principle ::= using @one_term {? with @bindings } {? @occurrences }
 
    Performs case analysis by generating a subgoal for each constructor of the
    inductive or co-inductive type in the :n:`@one_term` in :n:`@induction_arg`.
+   :n:`@one_term` (or the conclusion of its type--example
+   :ref:`here<example_destruct_ind_concl>`) must be of inductive or co-inductive type.
    In each new subgoal, the tactic replaces :n:`@one_term` with the associated
    constructor (applied to its arguments).   Unlike :tacn:`induction`,
    :tacn:`destruct` generates no induction hypothesis.
```
Let's change the wording entirely to make this clearer. Something along these lines:

```suggestion
   The main argument must be a term or a premise whose type is an inductive or
   co-inductive type (or a product type whose conclusion is an inductive or
   co-inductive type). This tactic performs case analysis by generating a subgoal
   for each constructor of the inductive or co-inductive type. Unlike
   :tacn:`induction`, :tacn:`destruct` generates no induction hypothesis.
```

Then, you could insert the bullet-list explaining the various ways that the main argument can be provided.

jfehrle replied:
OK, surgery complete.  Let me know what you think.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -189,6 +190,25 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
 
+.. _example_destruct_ind_concl:
+
+   .. example:: :tacn:`destruct` on a type with a conclusion that is an inductive type
```
A less complex title would be:
```suggestion
   .. example:: :tacn:`destruct` called on an argument with premises
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -260,8 +281,7 @@ Induction
 
-   Most parameters behave as they do in :tacn:`destruct`, with the following
-   differences:
+   Most parameters behave as they do in :tacn:`destruct`:
```
Right. So now the question becomes if we want to repeat the information that's already supposed to be available in `destruct`. Up to you.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -442,140 +407,216 @@ Induction
+
+   Starts a proof by co-induction. The :n:`@ident`\s (including the first one
+   before the `with` clause) are the
+   names of the co-induction hypothesis. As in a cofix expression,
```
```suggestion
   names of the co-induction hypotheses. As in a cofix expression,
```

jfehrle replied:
Got it.  One of the more obscure ways to form a plural in English.

jfehrle commented:
Updated.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -127,16 +127,37 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
+   + If :n:`@one_term` (in :n:`@induction_arg`)
+     is an identifier :n:`@ident` denoting a quantified variable of the
+     goal, then :n:`destruct @ident` behaves like :tacn:`intros` :n:`until
+     @ident; destruct @ident`. :n:`@one_term` must have an inductive type.
```
This paragraph is a bit mixed up between the specific `ident` case and the generic `one_term` case. Since this last sentence applies in all cases, it could be factorized somewhere.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -127,16 +127,37 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
+     goal after application of :n:`destruct`, it is erased. To avoid erasure,
+     use parentheses, as in :n:`destruct (@ident)`.
+
+     The :n:`@one_term` may contain holes that are denoted by “_”. In this case,
```
Since this is a separate case from the `ident` case, it would deserve a separate bullet-point.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -127,16 +127,37 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
+     is an identifier :n:`@ident` denoting a quantified variable of the
+     goal, then :n:`destruct @ident` behaves like :tacn:`intros` :n:`until
+     @ident; destruct @ident`. :n:`@one_term` must have an inductive type.
+     If :n:`@ident` is no longer dependent in the
+     goal after application of :n:`destruct`, it is erased. To avoid erasure,
+     use parentheses, as in :n:`destruct (@ident)`.
```
This is said twice in case `ident` is a quantified variable of the goal or a hypothesis of the context. It would be worth factorizing this information in some unique place.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -146,10 +167,6 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
-
    :n:`@ident` (in :n:`@induction_arg`)
      If :token:`ident` is a quantified variable in the conclusion
      of the goal, then :n:`destruct @ident` behaves like :tacn:`intros`
```
This whole thing is now redundant with the bullet points above.

jfehrle commented:
Updated.  I got your changes (I think) and reordered the destruct description a bit.  I also dropped some `tacn` hyperlinks--it seems pointless for the text in the description of, say, `destruct` hyperlink back to itself (though that will show it in black rather than blue).

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -119,45 +119,45 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
      Giving multiple :n:`@induction_clause`\s is equivalent to applying :n:`destruct`
      serially on each :n:`@induction_clause`.
 
+   :n:`@induction_arg`
+     + If :n:`@one_term` (in :n:`@one_term_with_bindings`)
+       is an identifier :n:`@ident` denoting a quantified variable of the
+       goal, then :n:`destruct @ident` behaves like :tacn:`intros` :n:`until
+       @ident; destruct @ident`. :n:`@one_term` must have an inductive type.
+       If :n:`@ident` is no longer dependent in the
+       goal after application of :n:`destruct`, it is erased. To avoid erasure,
+       use parentheses, as in :n:`destruct (@ident)`.
+     + :n:`@one_term` may contain holes that are denoted by “_”. In this case,
+       the tactic selects the first subterm that matches the pattern and performs
+       case analysis using that subterm.
+     + If :n:`@induction_arg` is a :n:`@natural`, then :n:`destruct @natural` behaves like
+       :n:`intros until @natural` followed by :n:`destruct` applied to the last
+       introduced hypothesis.
+
    :n:`@natural`
      Equivalent to :tacn:`intros` :n:`until @natural` followed by `destruct` applied to
      the last introduced hypothesis.
```
Note that this suggestion encompasses several changes:

- the note on `ident` being no longer dependent in the goal is not specific to the case where `ident` is a quantified variable of the goal
- `one_term` must have an inductive type was redundant with what is already written above and didn't belong in the middle of a bullet-point about `ident`
- the syntax case on `natural` was redundant with the corresponding bullet-point.

```suggestion
   :n:`@induction_arg`
     + If :n:`@one_term` (in :n:`@one_term_with_bindings`)
       is an identifier :n:`@ident`:
       + If :n:`@ident` denotes a quantified variable of the
         goal, then :n:`destruct @ident` behaves like :tacn:`intros` :n:`until
         @ident; destruct @ident`.
       + If :n:`@ident` is no longer dependent in the
         goal after application of :n:`destruct`, it is erased. To avoid erasure,
         use parentheses, as in :n:`destruct (@ident)`.
     + :n:`@one_term` may contain holes that are denoted by “_”. In this case,
       the tactic selects the first subterm that matches the pattern and performs
       case analysis using that subterm.
     + If :n:`@induction_arg` is a :n:`@natural`, then :n:`destruct @natural` behaves like
       :n:`intros until @natural` followed by :n:`destruct` applied to the last
       introduced hypothesis.
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,202 +256,114 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
 
-.. tacv:: elim @term using @term
-          elim @term using @term with @bindings
+  :n:`with @bindings`   (in :n:`@one_term_with_bindings`)
+    Explicitly gives instances to the premises of the type of :n:`@one_term`
+    (see :ref:`bindings`).
```
```suggestion
   :n:`with @bindings`   (in :n:`@one_term_with_bindings`)
     Explicitly gives instances to the premises of the type of :n:`@one_term`
     (see :ref:`bindings`).
```

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -237,202 +256,114 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
-   is not the standard one for the underlying inductive type of :n:`@term`. The
-   :n:`@bindings` clause allows instantiating premises of the type of
-   :n:`@term`.
+  :n:`{? using @one_term {? with @bindings } }`
+    Allows explicitly giving an induction principle :n:`@one_term` that
+    is not the standard one for the underlying inductive type of :n:`@one_term`. The
+    :n:`@bindings` clause allows instantiating premises of the type of
+    :n:`@one_term`.
```
```suggestion
   :n:`{? using @one_term {? with @bindings } }`
     Allows explicitly giving an induction principle :n:`@one_term` that
     is not the standard one for the underlying inductive type of :n:`@one_term`. The
     :n:`@bindings` clause allows instantiating premises of the type of
     :n:`@one_term`.
```

jfehrle commented:
Updated.  Anything else before I squash and filter out changes after line ~500?

Zimmi48 reviewed (approved):
Nothing else :smile:

Zimmi48 commented:
It's been a pleasure reviewing the changes to this chapter. The strategy to review the changes to a file in different chunks has worked out really well!

jfehrle commented:
Updated, should be good to merge.  I trimmed everything after the description of `cofix` and disabled `doc_grammar` from processing `reasoning-inductives.rst`.

Zimmi48 commented on doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:
```diff
@@ -189,6 +189,25 @@ analysis on inductive or co-inductive objects (see :ref:`variants`).
      Makes the tactic equivalent to
      :tacn:`induction` :n:`{+, @induction_clause } @induction_principle`.
 
+.. _example_destruct_ind_concl:
```
```suggestion
   .. _example_destruct_ind_concl:
```

jfehrle replied:
Yes, this was the cause.

Zimmi48 reviewed (commented):
Sorry, while throwing a last look at the rendered manual, I noticed something strange with the color of the `edestruct` tactic (blue instead of gray). My hypothesis is that this is because of this label that is incorrectly indented and breaks the description block of `destruct`.

Zimmi48 commented:
Something else that is incorrect after your last update is that you forgot to remove the former documentation of `fix` / `cofix` (after `inversion_sigma`).

jfehrle commented:
I removed the extra fix and cofix.  I'm surprised Sphinx didn't complain.

Updated.

Zimmi48 commented:
@coqbot merge now

jfehrle commented:
Cool, thanks!  Isn't is past your bedtime?

Zimmi48 commented:
Usually yes.
